### ISSUE TITLE
How to get the window and document object?

### ISSUE THREAD

**Issue by marcorra** [babel/babel-library-boilerplate#148]
Hi

Seems to be a great boilerplate, but I got an issue where I can't get the window and document object. Whatever I try, seems to be wrong.

Example a simple line like:  document.createElement("div");

is impossible. Is there any ways to get this to work?

It seems to work just fine when you $ gulp build,  but on testing it throws and let you know this is not defined.

However, if you just build and skip the testing and try the code it all work.

So very interested in how to solve this.

**Comment by jmeas** [comment]
Can you post the error?

**Comment by marcorra** [comment]
Well, I don't do anything then have a simple javascript file like:

export const WINDOW = window;
export const DOCUMENT = document;

so if you or me at least running $ gulp test, I get this:

Message:
    window is not defined
Stack:
ReferenceError: window is not defined
    at Object.<anonymous> (E:/xampp12/htdocs/src/environment.js:7:23)
    at Module._compile (module.js:460:26)
    at normalLoader (E:\xampp12\htdocs\node_modules\babel\node_modules\babel-core\lib\babel\api\register\node.js:147:5)
    at Object.require.extensions.(anonymous function) [as .js] (E:\xampp12\htdocs\node_modules\babel\node_modules\babel-core\lib\babel\api\register\node.js:160:7)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)

**Comment by marcorra** [comment]
Also. I tried several options. I know in other setups and both for ES6 transpiler and Babel you can set window and document to true in the lint setup files.

But in this boilerplate this also throws an errors.

Else this also throws:   /* globals window, document */

Whitelisting is not an option.

**Comment by marcorra** [comment]
Now I tried to add this into the jshintrc file:

  "globals": {
   "document": true,
    "window": true,
    "console": true
  }
and then grab this from the .js file

/* globals window, document */

But a dead-end this too.

I think I wasted 1 hour to much on this, so the solution would be to skip this boilerplate I guess.

**Comment by jmeas** [comment]
@marcorra, running `gulp test` runs the tests in a node environment, so there's no `window` or `document`. You need to use JSDom or Phantom to use the browser env variables.

**Comment by marcorra** [comment]
@jmeas Explain this if you can, because everything is tried here. If you can't well I find another boilerplate.

**Comment by jmeas** [comment]
@marcorra, JavaScript can be run in different environments. In the browser, you have access to variables like `window` and `document`. In Node, these objects don't exist. The problems you are running into have nothing to do with this boilerplate in particular; rather, it has to do with the fact that this boilerplate runs the tests through node when you run `gulp` or `gulp test`.

To access the browser variables in node, you have a few options. One is to run a headless browser, like PhantomJS. The other is to simulate the DOM, which is what JSDom provides.

Here's [an example library that sets up JSDom](https://github.com/jmeas/backbone.base-router/blob/master/test/setup/node.js#L2-L8).

To run the tests in the browser, follow the instructions in [the readme](https://github.com/babel/babel-library-boilerplate#browser-tests).

It's important that the tests run in a node environment, because it's convenient for setting up services like TravisCI.

Does that make sense?

**Comment by marcorra** [comment]
@jmeas 

Everything make sence to me, and I know a little about this things. Problem I'm getting is to include the needed stuff with this boilerplate. 

I'm not 100% sure how to fix that with the boilerplate. So if you can help?

JSDOM as you pointing out have no support for nodeJS after v. 4.0, only ios.

So I think the solution here would be to set up PhantomJS, but my attempt so far
to attach

- phantom.js
- mocha-phantomjs
- gulp-mocha-phantomjs

seems to take me now where.

It has to be run with the TravisCI. That for sure, but just now I face trouble with gulp.js file to set things up, and also in the /test/setup folder.

I'm not even sure how to get the TRAVIS_JOB_NUMBER number to work with this.

So I think I will stick with my old solution.

With my old solution however I modified karma.conf ( used Karma ) and in the Gulp fle i used this simple code:

 var config = {preprocessors: []};

    if (process.env.TRAVIS_JOB_NUMBER) {
        config = {
            reporters: ["coverage", "dots", "coveralls"],
            coverageReporter: {
                type: "lcovonly",
                dir: "coverage/"
            }
        };
    } else {
        if (argv.all) {
            config.browsers = ["PhantomJS", "Chrome", "ChromeCanary", "Opera", "Safari", "Firefox"];
        } else if (argv.ie8) {
            config.browsers = ["IE8 - WinXP"];
        } else if (argv.ie9 || argv.ie10 || argv.ie11) {
            config.browsers = ["IE" + (argv.ie9 ? "9" : (argv.ie10 ? "10" : "11")) + " - Win7"];
        }
    }

All this seems not to be possible with this boilerplate.

So if you have a good solution on setting this things up, I would be grateful. Else I think I skip this boilerplate.

**Comment by jmeas** [comment]
@marcorra, you should look on StackOverflow for more help on setting up Phantom or JSDom. It's beyond the scope of this repo to include a guide for those sorts of things.

**Comment by marcorra** [comment]
@jmeas Well. I have set up this things many times in my life. Problem was with this boilerplate. 

I just pushed the delete button now and removed this boilerplate, so case closed. Couldn't waste more time.

**Comment by marcorra** [comment]
@jmeas Did a clean install of this boilerplate because I got very angry now after 5 timers in hell with this shit, and then out of the blue I recived this error message:


Error loading resource file:///E:/xampp12/htdocs/node_modules/babel/browser-polyfill.js (203). Details: Error opening E:/xampp12/htdocs/node_modules/babel/browser-polyfill.js: No such file or directory
Error loading resource http://localhost:35729/livereload.js (1). Details: Connection refused

I didn't add anything, just tried to run the browser test

But I'm out of this now. Terrible, terrible setup.
